### PR TITLE
PDF Database

### PR DESCRIPTION
**Pull request by cameronfabbri**
This example shows how to populate a chromadb database with PDFs, then the ability to ask questions about the data.

You'll need to install `requirements.txt` again

### To run
First, download the Title-12 data
`python -m scripts.download_title_12`

This will save it in the directory `{projects.DATASETS_DIR_PATH}/Title-12/`

Because the PDFs are very large, I suggest creating another directory to test called `Title-12-Small`, e.g., if `projects.DATASETS_DIR_PATH = data/`
`mkdir data/Title-12-Small`

then extracting a range of pages:
`python -m scripts.extract_pdf --file_in data/Title-12/Title-12-Volume-1.pdf --file_out data/Title-12-Small/Title-12-Volume-1.pdf --start 16 --end 31`

You can then run the example which will spin up gradio
`python -m examples.pdf_chat --data_dir data/Title-12-Small`

- [x] Class for a chromadb PDF database
- [x] Store embeddings of full PDF in database
- [x] Store embeddings of individual pages in database
- [x] Store embeddings of partial pages in database
- [x] Search database by filename
- [x] Search database by comparing embeddings

### CODE REVIEW

**Comment by jmpounders** [comment]
Looks like there is a missing module, `text_utils`

```
> python -m examples.pdf_chat --data_dir data/Title-12-Small
found config `MODELS_DIR_PATH` in config file
found config `DATASETS_DIR_PATH` in config file
found config `GENERATOR_CACHE_DIR_PATH` in config file
found config `OPENAI_API_KEY_FILE_PATH` in config file
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/jmpounders/neo-sophia/examples/pdf_chat.py", line 22, in <module>
    from neosophia.llmtools import openaiapi as oaiapi, pdf_utils, text_utils
ImportError: cannot import name 'text_utils' from 'neosophia.llmtools' (unknown location)
```

**Comment by jmpounders** [comment]
Nice work.  I think this sets us up nicely to use PDF search as an LLM tool.  The PDFDB and related parsing utils are really useful.

I left a few minor comments.

File: examples/pdf_chat.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+"""
```
Comment (bdzimmer) [inline examples/pdf_chat.py:1]: Add docstring here and in other modules

File: examples/pdf_chat.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    # Create database
+    pdfdb = PDFDB('pdf_db', api_key)
+
+    # Add PDFs from the directory to the database
+    filenames = sorted(pdf_utils.find_pdfs_in_directory(data_dir))
+    for filename in tqdm(filenames):
+        pdfdb.add_pdf(filename)
+
+    base_prompt = 'Determine if the following context has enough information to answer the question. If it doesn\'t, say "There is not enough context to provide an answer". If it does, provide the answer. THINK STEP BY STEP. ALWAYS PROVIDE QUOTES AND PAGE CITATIONS. ALWAYS PROVIDE AN ESTIMATED CONFIDENCE SCORE WITH YOUR ANSWER\n\n'
```
Comment (bdzimmer) [inline examples/pdf_chat.py:43]: Wrap here and below on line 70

File: examples/pdf_chat.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    def ask_question(question):
```
Comment (bdzimmer) [inline examples/pdf_chat.py:45]:
```suggestion
    def ask_question(question: str) -> str:
```

File: examples/pdf_chat.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
+"""
+"""
+import os
+import re
+import time
+import pickle
```
Comment (bdzimmer) [inline examples/pdf_chat.py:6]: Remove extra imports

File: scripts/download_title_12.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+""" Script to download Volumes 1-10 of Title 12 """
+import os
+
+import click
+import requests
+
+from tqdm import tqdm
```
Comment (bdzimmer) [inline scripts/download_title_12.py:7]: Remove unused import

File: src/neosophia/llmtools/text_utils.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+""" Text utilities """
+from neosophia.llmtools import openaiapi as oaiapi
```
Comment (bdzimmer) [inline src/neosophia/llmtools/text_utils.py:2]: remove

**Review by bdzimmer — APPROVED** [review]

**Pull request merged**